cajita/src/Cajita_SparseLocalGrid_impl.hpp

```diff
@@ -0,0 +1,434 @@
+template <class Scalar, std::size_t NumSpaceDim>
+int LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::haloTileWidth() const
+{
+    return _halo_cell_width / _cell_num_per_tile_dim;
+}
+
+//---------------------------------------------------------------------------//
+//! Get the total number of local cells in a given dimension (owned + halo).
+template <class Scalar, std::size_t NumSpaceDim>
+int LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::totalNumCell(
+    const int d ) const
+{
+    return _global_grid->ownedNumCell( d ) + 2 * _halo_cell_width;
+}
+
+//---------------------------------------------------------------------------//
+// Get the total number of local cells in current rank (owned + halo).
+template <class Scalar, std::size_t NumSpaceDim>
+int LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::totalNumCell() const
+{
+    int total_num_cell = 1;
+    for ( std::size_t d = 0; d < num_space_dim; ++d )
+        total_num_cell *= totalNumCell( static_cast<int>( d ) );
+
+    return total_num_cell;
+}
+
+//---------------------------------------------------------------------------//
+//! Given the relative offsets of a neighbor rank relative to this local grid's
+//! indices get the of the neighbor.
+template <class Scalar, std::size_t NumSpaceDim>
+int LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::neighborRank(
+    const std::array<int, num_space_dim>& off_ijk ) const
+{
+    std::array<int, num_space_dim> nijk;
+    for ( std::size_t d = 0; d < num_space_dim; ++d )
+        nijk[d] = _global_grid->dimBlockId( d ) + off_ijk[d];
+    return _global_grid->blockRank( nijk );
+}
+
+template <class Scalar, std::size_t NumSpaceDim>
+int LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::neighborRank(
+    const int off_i, const int off_j, const int off_k ) const
+{
+    std::array<int, num_space_dim> off_ijk = { off_i, off_j, off_k };
+    return neighborRank( off_ijk );
+}
+
+//---------------------------------------------------------------------------//
+//! Get the index space for a given combination of decomposition, entity, and
+//! index types.
+template <class Scalar, std::size_t NumSpaceDim>
+template <class DecompositionTag, class EntityType, class IndexType>
+auto LocalGrid<SparseMesh<Scalar, NumSpaceDim>>::indexSpace(
+    DecompositionTag t1, EntityType t2, IndexType t3 ) const
+    -> IndexSpace<num_space_dim>
+{
+    return indexSpaceImpl( t1, t2, t3 );
+}
+
+//---------------------------------------------------------------------------//
+/*!
+    \brief Given the relative offsets of a neighbor rank relative to this local
+    grid's indices get the set of local entity indices shared with that
+    neighbor in the given decomposition. Optionally provide a halo width
+    for the shared space. This halo width must be less than or equal to the
+    halo width of the local grid. The default behavior is to use the halo
+    width of the local grid.
+    \param t1 Decomposition type: Own or Ghost
+    \param t2 Entity: Cell or Node; TODO: Edge or Face
```

Comment:
This will already error for edge or face, right?

Reply:
Yeah, it should be.
I mean, here, Cell and Node are supported, but Edge and Face are still on the to-do list.

Reply:
+comment: why do cell and node share the same interface; and "we can add support for edge/face later"; uniqueness thing